### PR TITLE
SALTO-6211: add fixer for merge of options lists in ticket_field

### PR DESCRIPTION
**Pull request by edenhassid**
add fixer for merge of  options lists in ticket_field

---

_Additional context for reviewer_
In this fixer we merge custom_field_options, by updating the father list to include all options from the elementsSource.
Therefor if there is a custom option that was deleted from the list however its instance was not deleted we will add it back to the list

---
_Release Notes_: 
_Replace me with a short sentence that describes the effect of this change on Salto users_

---
_User Notifications_: 
_Replace me with a short sentence that describes changes that will appear in NaCls and are not caused by user actions (e.g. a new annotation, field values that are converted to references, etc). Hidden changes should not be listed._

### CODE REVIEW

File: packages/zendesk-adapter/src/config.ts [hunk 1]
```diff
@@ -3104,6 +3105,7 @@ export const configType = createMatchingObjectType<Partial<ZendeskConfig>>({
       refType: definitions.createUserDeployConfigType(ZENDESK, changeValidatorConfigType, {
         ...defaultMissingUserFallbackField,
         createMissingOrganizations: { refType: BuiltinTypes.BOOLEAN },
+        fixParentOption: { refType: BuiltinTypes.BOOLEAN },
```
Comment (ori-moisis) [inline packages/zendesk-adapter/src/config.ts:3108]: the configuration should be under the same key for all fixers, and then the enabled fixers should be generic (should pass as a second parameter to `combineElementFixers`)

File: packages/zendesk-adapter/src/fix_elements/merge_lists.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+import { collections, values as lowerDashValues } from '@salto-io/lowerdash'
+import { FixElementsHandler } from './types'
+import { TICKET_FIELD_CUSTOM_FIELD_OPTION, TICKET_FIELD_TYPE_NAME } from '../constants'
+
+const { isDefined } = lowerDashValues
+const { awu } = collections.asynciterable
+const log = logger(module)
+
+const PARENT_TO_CHILD_MAP: Record<string, string> = {
+  [TICKET_FIELD_TYPE_NAME]: TICKET_FIELD_CUSTOM_FIELD_OPTION,
```
[the remark this answers is not in the document]
Comment (edenhassid) [reply packages/zendesk-adapter/src/fix_elements/merge_lists.ts:39]: I thought that we could add  user and org fields when we encounter this issue for them

File: packages/zendesk-adapter/src/fix_elements/merge_lists.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+      })
+      .filter(isDefined)
+      .map(parent => parent.elemID.getFullName())
+      // we need to get all the parent names that don't appear already in the elements
+      .filter(name => !parentNames.has(name)),
+  )
+  const parentElementsFromElementSource: InstanceElement[] = await awu(await elementsSource.list())
+    .filter(id => id.idType === 'instance' && parentNamesFromChild.has(id.getFullName()))
+    .map(id => elementsSource.get(id))
+    .toArray()
```
Comment (ori-moisis) [inline packages/zendesk-adapter/src/fix_elements/merge_lists.ts:75]: we already have the exact ids, we don't need the `list` + `filter`

File: packages/zendesk-adapter/src/fix_elements/merge_lists.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+const getChildrenElemIdByParent = (allParents: InstanceElement[], elemIdsList: ElemID[]): Record<string, ElemID[]> =>
+  Object.fromEntries(
+    allParents.map(parent => {
+      const parentElemId = parent.elemID
+      const childrenElemIds = elemIdsList.filter(id =>
+        id
+          .getFullName()
+          .startsWith(
+            `${parentElemId.adapter}.${PARENT_TO_CHILD_MAP[parentElemId.typeName] ?? ''}.instance.${pathNaclCase(parentElemId.name)}`,
```
[the remark this answers is not in the document]
Comment (edenhassid) [reply packages/zendesk-adapter/src/fix_elements/merge_lists.ts:87]: its correct for user and org field as well

File: packages/zendesk-adapter/src/fix_elements/merge_lists.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  )
+
+type FixedParentResult = { fixedParent: InstanceElement; childrenAdded: string[] }
+
+const fixParent = (
+  parent: InstanceElement,
+  childrenElemIdByParent: Record<string, ElemID[]>,
+  allChildrenByFullName: Record<string, InstanceElement>,
+): FixedParentResult | undefined => {
+  const parentOptions = parent.value.custom_field_options
```
[the remark this answers is not in the document]
Comment (edenhassid) [reply packages/zendesk-adapter/src/fix_elements/merge_lists.ts:101]: user, ticket, and org field all use the same `custom_field_options` if we would add other types in the future we can configure that. i'll add a comment

File: packages/zendesk-adapter/test/fix_elements/merge_lists.test.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+      new ReferenceExpression(ticketFiledOption2.elemID, ticketFiledOption2),
+    ]
+    config = {
+      ...DEFAULT_CONFIG,
+      [DEPLOY_CONFIG]: { fixParentOption: true },
+    }
+    client = new ZendeskClient({
+      credentials: { username: 'a', password: 'b', subdomain: 'ignore' },
+    })
+    elementsSource = elementSource.createInMemoryElementSource([clonedTicket, ticketFiledOption1, ticketFiledOption2])
```
[the remark this answers is not in the document]
Comment (edenhassid) [reply packages/zendesk-adapter/test/fix_elements/merge_lists.test.ts:62]: no, the cloned one is based on the original ticket, we change its options each test, this is why the cloned one should be in the elementSource

File: packages/zendesk-adapter/src/fix_elements/merge_lists.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  if (!_.isArray(parentOptions)) {
```
Comment (ori-moisis) [inline packages/zendesk-adapter/src/fix_elements/merge_lists.ts:102]: warning log?

File: packages/zendesk-adapter/test/fix_elements/merge_lists.test.ts [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  let clonedTicket: InstanceElement
+  let fixedTicket: InstanceElement
+  const ticketFieldType = new ObjectType({ elemID: new ElemID(ZENDESK, TICKET_FIELD_TYPE_NAME) })
+  const customFieldOptionType = new ObjectType({ elemID: new ElemID(ZENDESK, TICKET_FIELD_CUSTOM_FIELD_OPTION) })
+
+  const ticketField1 = new InstanceElement('ticket1', ticketFieldType, {})
+  const ticketFiledOption1 = new InstanceElement('ticket1Option1', customFieldOptionType, {}, undefined, {
+    [CORE_ANNOTATIONS.PARENT]: [new ReferenceExpression(ticketField1.elemID, ticketField1)],
+  })
+  const ticketFiledOption2 = new InstanceElement('ticket1Option2', customFieldOptionType, {}, undefined, {
```
Comment (adfineberg) [inline packages/zendesk-adapter/test/fix_elements/merge_lists.test.ts:44]:
```suggestion
  const ticketFieldOption1 = new InstanceElement('ticket1Option1', customFieldOptionType, {}, undefined, {
    [CORE_ANNOTATIONS.PARENT]: [new ReferenceExpression(ticketField1.elemID, ticketField1)],
  })
  const ticketFieldOption2 = new InstanceElement('ticket1Option2', customFieldOptionType, {}, undefined, {
```

File: packages/zendesk-adapter/test/fix_elements/merge_lists.test.ts [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    expect(fixerRes.errors.length).toEqual(1)
+    expect(fixerRes.errors[0]).toEqual({
+      elemID: fixerRes.fixedElements[0].elemID,
+      severity: 'Warning',
+      message: 'custom_field_options were updated',
+      detailedMessage:
+        'ticket_field custom_field_options were updated.\nThe following options were added at the end of the list: zendesk.ticket_field__custom_field_options.instance.ticket1Option2.',
+    })
+  })
+  it('Should add remove options from the parent list when it does not appear in the elementsSource', async () => {
```
Comment (adfineberg) [inline packages/zendesk-adapter/test/fix_elements/merge_lists.test.ts:84]:
```suggestion
  it('Should remove options from the parent list when it does not appear in the elementsSource', async () => {
```

File: packages/zendesk-adapter/src/adapter.ts [hunk 10]
```diff
@@ -556,6 +556,7 @@ export default class ZendeskAdapter implements AdapterOperations {
         config,
         elementsSource,
       }),
+      config.fixElements ? (config.fixElements as Record<string, boolean>) : undefined,
```
Comment (adfineberg) [inline packages/zendesk-adapter/src/adapter.ts:559]: If there's a good way to remove this casting, let's do it. Not a blocker for approval IMO

File: packages/zendesk-adapter/src/fix_elements/merge_lists.ts [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+  const childrenAddedMsg = !_.isEmpty(fixedParentsRes.childrenAdded)
+    ? `\nThe following options were added at the end of the list: ${fixedParentsRes.childrenAdded.join(', ')}.`
+    : ''
+  const childrenRemovedMsg = !_.isEmpty(fixedParentsRes.childrenRemoved)
+    ? `\nThe following options were removed from the list: ${fixedParentsRes.childrenRemoved.join(', ')}.`
+    : ''
+  return {
+    elemID: fixedParentsRes.fixedParent.elemID,
+    severity: 'Warning',
+    message: 'custom_field_options were updated',
```
Comment (adfineberg) [inline packages/zendesk-adapter/src/fix_elements/merge_lists.ts:166]: Let's run this through @tomermevorach for wording

**Review by adfineberg — APPROVED** [review]
Some small fixes, pre approving 💯

**Pull request merged**